So, here's what mypy complains about on my computer:

```console
Tools/clinic/clinic.py:536: error: Incompatible types in assignment (expression has type "list[Iterable[Parameter] | Parameter]", variable has type "list[Iterable[Parameter]]")  [assignment]
Tools/clinic/clinic.py:553: error: Argument 1 to "extend" of "list" has incompatible type "Iterable[Parameter]"; expected "Iterable[Iterable[Parameter]]"  [arg-type]
Tools/clinic/clinic.py:584: error: Argument 1 to "append" of "list" has incompatible type "tuple[Iterable[Parameter], ...]"; expected "Iterable[Parameter]"  [arg-type]
Tools/clinic/clinic.py:586: error: Argument "key" to "sort" of "list" has incompatible type "Callable[[Sized], int]"; expected "Callable[[Iterable[Parameter]], SupportsDunderLT[Any] | SupportsDunderGT[Any]]"  [arg-type]
```

Is mypy fooled by the `list(group)` thing? Also, I'm puzzled by the `key` param complaint.

--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -28,7 +28,12 @@ import sys
 import textwrap
 import traceback
 
-from collections.abc import Callable, Generator, Iterable, Reversible
+from collections.abc import (
+    Callable,
+    Iterable,
+    Iterator,
+    Sequence,
+)
 from types import FunctionType, NoneType
 from typing import (
     Any,
@@ -516,13 +521,12 @@ class PythonLanguage(Language):
     checksum_line = "#/*[{dsl_name} end generated code: {arguments}]*/"
 
 
-ParamGroup = Iterable["Parameter"]
-ParamGenerator = Generator[tuple[ParamGroup, ...], None, None]
+ParamTuple = tuple["Parameter", ...]
 
 
 def permute_left_option_groups(
-        l: Reversible[ParamGroup]
-) -> ParamGenerator:
+        l: Sequence[ParamTuple]
+) -> Iterator[ParamTuple]:
     """
     Given [(1,), (2,), (3,)], should yield:
        ()
@@ -531,15 +535,15 @@ def permute_left_option_groups(
        (1, 2, 3)
     """
     yield tuple()
-    accumulator: list[ParamGroup] = []
+    accumulator: list[Parameter] = []
     for group in reversed(l):
         accumulator = list(group) + accumulator
         yield tuple(accumulator)
 
 
 def permute_right_option_groups(
-        l: Iterable[ParamGroup]
-) -> ParamGenerator:
+        l: Sequence[ParamTuple]
+) -> Iterator[ParamTuple]:
     """
     Given [(1,), (2,), (3,)], should yield:
       ()
@@ -548,17 +552,17 @@ def permute_right_option_groups(
       (1, 2, 3)
     """
     yield tuple()
-    accumulator: list[ParamGroup] = []
+    accumulator: list[Parameter] = []
     for group in l:
         accumulator.extend(group)
         yield tuple(accumulator)
 
 
 def permute_optional_groups(
-        left: Reversible[ParamGroup],
-        required: Iterable[ParamGroup],
-        right: Iterable[ParamGroup]
-) -> tuple[ParamGroup, ...]:
+        left: Sequence[ParamTuple],
+        required: ParamTuple,
+        right: Sequence[ParamTuple]
+) -> tuple[ParamTuple, ...]:
     """
     Generator function that computes the set of acceptable
     argument lists for the provided iterables of
@@ -573,7 +577,7 @@ def permute_optional_groups(
         if left:
             raise ValueError("required is empty but left is not")
 
-    accumulator: list[ParamGroup] = []
+    accumulator: list[ParamTuple] = []
     counts = set()
     for r in permute_right_option_groups(right):
         for l in permute_left_option_groups(left):

--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -525,7 +525,7 @@ ParamTuple = tuple["Parameter", ...]
 
 
 def permute_left_option_groups(
-        l: Sequence[ParamTuple]
+        l: Sequence[Iterable[Parameter]]
 ) -> Iterator[ParamTuple]:
     """
     Given [(1,), (2,), (3,)], should yield:
@@ -542,7 +542,7 @@ def permute_left_option_groups(
 
 
 def permute_right_option_groups(
-        l: Sequence[ParamTuple]
+        l: Sequence[Iterable[Parameter]]
 ) -> Iterator[ParamTuple]:
     """
     Given [(1,), (2,), (3,)], should yield:
@@ -559,9 +559,9 @@ def permute_right_option_groups(
 
 
 def permute_optional_groups(
-        left: Sequence[ParamTuple],
-        required: ParamTuple,
-        right: Sequence[ParamTuple]
+        left: Sequence[Iterable[Parameter]],
+        required: Iterable[Parameter],
+        right: Sequence[Iterable[Parameter]]
 ) -> tuple[ParamTuple, ...]:
     """
     Generator function that computes the set of acceptable

